test/performance/odp_timer_perf.c

```diff
@@ -553,6 +621,143 @@ static int test_worker(void *arg)
 	return ret;
 }
 
+static int set_cancel_mode_worker(void *arg)
+{
+	uint64_t tick, start_tick, period_tick, nsec;
+	uint64_t c1, c2, diff;
+	int thr, status;
+	uint32_t i, j, worker_idx;
+	odp_event_t ev;
+	odp_time_t t1, t2;
+	odp_timer_t timer;
+	odp_timer_pool_t tp;
+	odp_timeout_t tmo;
+	thread_arg_t *thread_arg = arg;
+	test_global_t *global = thread_arg->global;
+	test_options_t *test_options = &global->test_options;
+	uint32_t num_tp = test_options->num_tp;
+	uint32_t num_timer = test_options->num_timer;
+	uint32_t num_worker = test_options->num_cpu;
+	int ret = 0;
+	int started = 0;
+	uint64_t test_rounds = test_options->test_rounds;
+	uint64_t num_tmo = 0;
+	uint64_t num_cancel = 0;
+	uint64_t num_set = 0;
+
+	thr = odp_thread_id();
+	worker_idx = thread_arg->worker_idx;
+	t1 = ODP_TIME_NULL;
+	c1 = 0;
+
+	/* Start all workers at the same time */
+	odp_barrier_wait(&global->barrier);
+
+	while (1) {
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+
+		if (odp_unlikely(ev != ODP_EVENT_INVALID)) {
+			/* Timeout, set timer again. When start_tick is large enough, this should
+			 * not happen. */
+			timer_ctx_t *ctx;
+
+			tmo   = odp_timeout_from_event(ev);
+			ctx   = odp_timeout_user_ptr(tmo);
+			i     = ctx->tp_idx;
+			j     = ctx->timer_idx;
+			timer = global->timer[i][j];
+			start_tick  = global->timer_pool[i].start_tick;
+			period_tick = global->timer_pool[i].period_tick;
+			tick = start_tick + j * period_tick;
+
+			status = odp_timer_set_rel(timer, tick, &ev);
+			num_tmo++;
+			num_set++;
+
+			if (status != ODP_TIMER_SUCCESS) {
+				printf("Error: Timer set (tmo) failed (ret %i)\n", status);
+				ret = -1;
+				break;
+			}
+
+			continue;
+		}
+
+		if (odp_unlikely(odp_atomic_load_u32(&global->exit_test)))
+			break;
+
+		if (odp_unlikely(started == 0)) {
+			/* Run schedule loop while waiting for timers to be created */
+			if (odp_atomic_load_acq_u32(&global->timers_started) == 0)
+				continue;
+
+			/* Start measurements */
+			started = 1;
+			t1 = odp_time_local();
+			c1 = odp_cpu_cycles();
+		}
+
+		/* Cancel and set timers again */
+		for (i = 0; i < num_tp; i++) {
+			tp = global->timer_pool[i].tp;
```

Comment:
If more than two timer pools are used all resources are not released correctly.

E.g.

```
$ sudo ./odp_timer_perf -m1 -c 1 -n 5
HW time counter freq: 2596984766 hz

Pool config:
  pool.local_cache_size: 256
  pool.burst_size: 32
  pool.pkt.max_num: 262143
  pool.pkt.base_align: 64
  pool.buf.min_align: 64

Queue config:
  queue_basic.max_queue_size: 8192
  queue_basic.default_queue_size: 4096

Using scheduler 'basic'
Scheduler config:
  sched_basic.prio_spread: 4
  sched_basic.prio_spread_weight: 63
  sched_basic.burst_size_default[] =  32  32  32  32  32  16   8   4
  sched_basic.burst_size_max[] =     255 255 255 255 255  16  16   8
  sched_basic.group_enable.all: 1
  sched_basic.group_enable.worker: 1
  sched_basic.group_enable.control: 1

Packet IO config:
  pktio.pktin_frame_offset: 0

PKTIO: initialized loop interface.
PKTIO: initialized dpdk pktio, use export ODP_PKTIO_DISABLE_DPDK=1 to disable.
PKTIO: initialized pcap interface.
PKTIO: initialized null interface.
PKTIO: initialized socket mmap, use export ODP_PKTIO_DISABLE_SOCKET_MMAP=1 to disable.
PKTIO: initialized socket mmsg,use export ODP_PKTIO_DISABLE_SOCKET_MMSG=1 to disable.

ODP crypto is DISABLED

ODP compress is DISABLED

ODP traffic manager is DISABLED

ODP IPSec is DISABLED

ODP system info
---------------
ODP API version:  1.25.1
ODP impl name:    odp-linux
ODP impl details: odp-linux 1.25.1-0 (v1.25.1) 1.25.1.0
CPU model:        Intel(R) Xeon(R) CPU E5-2697 v3
CPU freq (hz):    2600000000
Cache line size:  64
CPU count:        27
CPU mask:         0xFFFFFFE


CPU features supported:
SSE3 PCLMULQDQ DTES64 MONITOR DS_CPL VMX SMX EIST TM2 SSSE3 FMA CMPXCHG16B XTPR PDCM PCID DCA SSE4_1 SSE4_2 X2APIC MOVBE POPCNT TSC_DEADLINE AES XSAVE OSXSAVE AVX F16C RDRAND FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE DIGTEMP ARAT PLN ECMD PTM MPERF_APERF_MSR ENERGY_EFF FSGSBASE AVX2 BMI2 ERMS INVPCID LAHF_SAHF SYSCALL XD 1GB_PG RDTSCP EM64T INVTSC 

CPU features NOT supported:
CNXT_ID PSN TRBOBST ACNT2 BMI1 HLE SMEP RTM AVX512F LZCNT 


Timer performance test
  mode             1
  num cpu          1
  private pool     0
  num timer pool   5
  num timer        10
  resolution       10000000 nsec
  period           100000000 nsec
  first timer at   0.10 sec
  test rounds      100000
  start            100000000 tick
  period           100000000 ticks

Timer pool info [0]:
  cur_timers       10
  hwm_timers       10

Timer pool info [1]:
  cur_timers       10
  hwm_timers       10

Timer pool info [2]:
  cur_timers       10
  hwm_timers       10

Timer pool info [3]:
  cur_timers       10
  hwm_timers       10

Timer pool info [4]:
  cur_timers       10
  hwm_timers       10


RESULTS - timer cancel + set cycles per thread:
-----------------------------------------------
        1      2      3      4      5      6      7      8      9     10
    289.6 

TOTAL (1 workers)
  rounds:              100000
  timeouts:            0
  timer cancels:       5000000
  timer sets:          5000000
  ave time:            0.56 sec
  cancel+set per cpu:  8.97M per sec

odp_queue_basic.c:418:queue_destroy():queue "" not empty
odp_queue_basic.c:418:queue_destroy():queue "" not empty
odp_queue_basic.c:418:queue_destroy():queue "" not empty
odp_schedule_basic.c:485:schedule_term_global():Queue not empty
odp_schedule_basic.c:485:schedule_term_global():Queue not empty
odp_schedule_basic.c:485:schedule_term_global():Queue not empty
odp_queue_basic.c:211:queue_term_global():Not destroyed queue: 
odp_queue_basic.c:211:queue_term_global():Not destroyed queue: 
odp_queue_basic.c:211:queue_term_global():Not destroyed queue: 
odp_init.c:202:term_global():ODP queue term failed.
Error: term global failed.
```

Reply:
fixed in v5: added timer cancel loop before worker thread exits